modules/versioned/data_test.go

```diff
@@ -0,0 +1,182 @@
+
+	_, err := NewReader(buf)
+	require.True(t, IsNotVersioned(err))
+
+	// case 2, invalid version string
+	buf = bytes.NewBufferString(`"abc" {"name": "Test", "value": "success"}`)
+
+	_, err = NewReader(buf)
+	require.True(t, IsNotVersioned(err))
+
+	// case 3, empty input
+	buf = bytes.NewBufferString("")
+	_, err = NewReader(buf)
+	require.True(t, IsNotVersioned(err))
+
+	// case 4, underlying io error
+	var bad BadReader
+	_, err = NewReader(bad)
+	require.Error(t, err)
+	require.False(t, IsNotVersioned(err))
+}
+
+func TestWriterReader(t *testing.T) {
+	type Data struct {
+		Name string
+		Age  float64
+		Tags []string
+	}
+
+	var buf bytes.Buffer
+
+	writer, err := NewWriter(&buf, MustParse("1.0.0"))
+	require.NoError(t, err)
+
+	// Note you can replace json here with any encoder u like
+	enc := json.NewEncoder(writer)
+	data := Data{"Test", 20.0, []string{"version", "check"}}
+	require.NoError(t, enc.Encode(data))
+
+	// loading
+	reader, err := NewReader(&buf)
+	require.NoError(t, err)
+
+	if ok := assert.Equal(t, MustParse("1.0.0"), reader.Version()); !ok {
+		t.Fatal()
+	}
+
+	dec := json.NewDecoder(reader)
+	var loaded Data
+
+	require.NoError(t, dec.Decode(&loaded))
+	if ok := assert.Equal(t, data, loaded); !ok {
+		t.Fatal()
+	}
+}
+
+func TestLoadSaveFile(t *testing.T) {
+	data := make([]byte, 500)
+	_, err := rand.Read(data)
+	require.NoError(t, err)
+
+	file, err := ioutil.TempFile("", "test-")
+	require.NoError(t, err)
+	// lazy way to get a temp file path
+	path := file.Name()
+	defer os.Remove(path)
+	file.Close()
+
+	version := MustParse("1.2.0-beta")
+	err = WriteFile(path, version, data, 0400)
+	require.NoError(t, err)
+
+	loadedVersion, loadedData, err := ReadFile(path)
+	require.NoError(t, err)
+
+	require.Equal(t, version, loadedVersion)
+	require.Equal(t, data, loadedData)
+}
+
+func TestLoadNotVersioned(t *testing.T) {
+	data := make([]byte, 500)
+	_, err := rand.Read(data)
+	require.NoError(t, err)
+
+	file, err := ioutil.TempFile("", "test-")
+	require.NoError(t, err)
+	// lazy way to get a temp file path
+	path := file.Name()
+	defer os.Remove(path)
+	_, err = file.Write(data)
+	require.NoError(t, err)
+	file.Close()
+
+	_, loadedData, err := ReadFile(path)
+	require.Error(t, err)
+	require.True(t, IsNotVersioned(err))
+
+	require.Equal(t, data, loadedData)
+}
+
+func ExampleNewReader() {
+	latest := MustParse("1.2.0")
+	// 1- Open file contains data
+	buf := bytes.NewBufferString(`"1.0.1" "my data goes here"`)
+
+	// 2- create versioned reader
+	reader, err := NewReader(buf)
+	if err != nil {
+		// no version in data, take another action!
+		panic(err)
+	}
+
+	fmt.Println("data version is:", reader.Version())
+	dec := json.NewDecoder(reader)
+
+	if reader.Version().Compare(latest) <= 0 {
+		//data version is older than or equal latest
+		var data string
+		if err := dec.Decode(&data); err != nil {
+			panic(err)
+		}
+		fmt.Println("data is:", "my data goes here")
+	}
```

Comment:
can we add an example how we would migrate the data if the version is not the one expected ?